heroku/deploy-a-django-app.md

```diff
@@ -0,0 +1,272 @@
+        - [`heroku.yml`](#heroku.yml)
+        - [`release.sh`](#release.sh)
+        - [`app.json`](#app.json)
+    - [Create apps and pipelines for your project](#create-apps-and-pipelines-for-your-project)
+- [Troubleshooting](#troubleshooting)
+
+## Set up application code for Heroku
+
+In order to deploy a Django application to Heroku, a few specific configurations
+need to be enabled in your application code. We provide details on these
+configurations below.
+
+### Containerize your app
+
+We use Heroku as a platform for deploying containerized apps, which means that
+your app must be containerized in order to use Heroku properly. If your app
+is not yet containerized, [follow our instructions for containerizing Django
+apps](/docker/local-development.md) before moving on.
+
+### Serve static files with WhiteNoise
+
+Apps deployed on Heroku don't typically use Nginx to serve content, so they need some
+other way of serving static files. Since our apps tend to have relatively low traffic,
+we prefer configuring [WhiteNoise](http://whitenoise.evans.io/)
+to allow Django to serve static files in production.
+
+Follow the [setup instructions for WhiteNoise in
+Django](http://whitenoise.evans.io/en/stable/django.html) to ensure that your
+Django apps can serve static files.
+
+In order to be able to serve static files when `DEBUG` is `False`, you'll also want
+to make sure that `RUN python manage.py collectstatic` is included as a step in your
+Dockerfile. This will ensure that the static files are baked into the container.
+Since we typically mount application code into the container during local development,
+you'll also need to make sure that your static files are stored outside of the root
+project folder so that the mounted files don't overwrite them. One easy way to do this
+is to set `STATIC_ROOT = '/static'` in your `settings.py` file.
+
+### Read settings and secret variables from the environment
+
+Heroku doesn't allow us to decrypt content with GPG, so we can't use Blackbox
+to decrypt application secrets. Instead, we can store these secrets as [config
+vars](https://devcenter.heroku.com/articles/config-vars), which Heroku will thread
+into our container environment.
+
+The three most basic config vars that you'll want to set for every app include
+the Django `DEBUG`, `SECRET_KEY`, and `ALLOWED_HOSTS` variables. Update `settings.py`
+to read these variables from the environment:
+
+```python
+SECRET_KEY = os.environ['DJANGO_SECRET_KEY']
+DEBUG = False if os.getenv('DJANGO_DEBUG', True) == 'False' else True
+allowed_hosts = os.getenv('DJANGO_ALLOWED_HOSTS', [])
+ALLOWED_HOSTS = allowed_hosts.split(',') if allowed_hosts else []
+```
+
+Make sure to update your app service in `docker-compose.yml` to thread any variables
+that don't have defaults into your local environment:
+
+```yml
+services:
+  app:
+    environment:
+      - DJANGO_SECRET_KEY=really-super-secret
+```
+
```

Comment:
Add a note to set the real values in the Heroku console before deploying (to save a round of builds that fail due to the missing variables).

Reply:
Looking at this again, I think it would be helpful to move this –

> Finally, configure environment variables for staging, production, and review apps. These can be set using either the dashboard or the CLI. Follow the Heroku documentation to set up your config vars.

– up two paragraphs, just after how to create an app but before instructions on configuring deployments. That way, if you're deploying from the branch you're working on (like I was), you don't have a failing deployment due to missing config values.

Reply:
Yeah, Fatima also got stuck here, so I think you're right that we should move it up. It also might be worth specifically calling out the config vars that this setup requires, including:

- `DJANGO_SECRET_KEY`
- `DJANGO_ALLOWED_HOSTS`

Reply:
Done in 115fe91603d94477f56ded5baf72a6d8abdb6926.